src/ligero/utils.rs

```diff
@@ -0,0 +1,325 @@
+        );
+
+        (0..self.m)
+            .map(|col| {
+                inner_product(
+                    v,
+                    &(0..self.n)
+                        .map(|row| self.entries[row][col])
+                        .collect::<Vec<F>>(),
+                )
+            })
+            .collect()
+    }
+}
+
+/// Compute the dimensions of an FFT-friendly (over F) matrix with at least n entries.
+/// The return pair (n, m) corresponds to the dimensions n x m.
+pub(crate) fn compute_dimensions<F: FftField>(n: usize) -> (usize, usize) {
+    assert_eq!(
+        (n as f64) as usize,
+        n,
+        "n cannot be converted to f64: aborting"
+    );
+
+    let aux = (n as f64).sqrt().ceil() as usize;
+    let n_cols = GeneralEvaluationDomain::<F>::new(aux)
+        .expect("Field F does not admit FFT with m elements")
+        .size();
+
+    (ceil_div(n, n_cols), n_cols)
+}
+
+/// Apply reed-solomon encoding to msg.
+/// Assumes msg.len() is equal to the order of an FFT domain in F.
+/// Returns a vector of length equal to the smallest FFT domain of size at least msg.len() * RHO_INV.
+pub(crate) fn reed_solomon<F: FftField>(
+    // msg, of length m, is interpreted as a vector of coefficients of a polynomial of degree m - 1
+    msg: &[F],
+    rho_inv: usize,
+) -> Vec<F> {
+    let m = msg.len();
+
+    let extended_domain = GeneralEvaluationDomain::<F>::new(m * rho_inv).unwrap_or_else(|| {
+        panic!(
+            "The field F cannot accomodate FFT for msg.len() * RHO_INV = {} elements (too many)",
+            m * rho_inv
+        )
+    });
+
+    extended_domain.fft(msg)
+}
+
+#[inline]
+pub(crate) fn inner_product<F: Field>(v1: &[F], v2: &[F]) -> F {
+    ark_std::cfg_iter!(v1)
+        .zip(v2)
+        .map(|(li, ri)| *li * ri)
+        .sum()
+}
+
+#[inline]
+#[cfg(test)]
+pub(crate) fn to_field<F: Field>(v: Vec<u64>) -> Vec<F> {
+    v.iter().map(|x| F::from(*x)).collect::<Vec<F>>()
+}
+
+#[inline]
+pub(crate) fn get_num_bytes(n: usize) -> usize {
+    ceil_div((usize::BITS - n.leading_zeros()) as usize, 8)
+}
+
+/// Takes as input a struct, and converts them to a series of bytes. All traits
+/// that implement `CanonicalSerialize` can be automatically converted to bytes
+/// in this manner.
+/// From jellyfish lib
+#[macro_export]
+macro_rules! to_bytes {
+    ($x:expr) => {{
+        let mut buf = ark_std::vec![];
+        ark_serialize::CanonicalSerialize::serialize_compressed($x, &mut buf).map(|_| buf)
+    }};
+}
+
+/// The following struct is taken from jellyfish repository. Once they change
+/// their dependency on `crypto-primitive`, we use their crate instead of
+/// a copy-paste. We needed the newer `crypto-primitive` for serializing.
+#[derive(Clone)]
+pub(crate) struct IOPTranscript<F: PrimeField> {
+    transcript: Transcript,
+    is_empty: bool,
+    #[doc(hidden)]
+    phantom: PhantomData<F>,
+}
+
+// TODO: merge this with jf_plonk::transcript
+impl<F: PrimeField> IOPTranscript<F> {
```

Comment:
This seems to me be yet another implementation of a sponge. We should try to consolidate all the usages so that we don't have duplication. cc @mmaker